### PR TITLE
Add new TCP input receiver.

### PR DESCRIPTION
**Pull request by jkschulz**
http://b/195571656

Manually tested and verified on Debian 10 (Buster).

#### 1. Build package locally for Debian 10/Buster

```
$ DOCKER_BUILDKIT=1 docker build --target buster -o /tmp/google-cloud-ops-agent .
```

#### 2. Create test VM and copy package over
```
$ gcloud compute instances create testtcp --image-project debian-cloud --image-family debian-10
$ gcloud compute scp /tmp/google-cloud-ops-agent/google-cloud-ops-agent_2.2.1~debian10_amd64.deb testtcp:/tmp/
```

#### 3. SSH to VM and install package
```
$ gcloud compute ssh testtcp
$ sudo apt install -y /tmp/google-cloud-ops-agent_2.2.1~debian10_amd64.deb
```

#### 4. Edit Ops Agent config and restart

```
jschulz@testtcp:~$ sudo vi /etc/google-cloud-ops-agent/config.yaml
jschulz@testtcp:~$ cat /etc/google-cloud-ops-agent/config.yaml
logging:
  receivers:
    tcp_logs:
      type: tcp
  service:
    pipelines:
      tcp_pipeline:
        receivers: [tcp_logs]
jschulz@testtcp:~$ sudo service google-cloud-ops-agent restart
```

#### 5. Confirm agent running successfully and generated Fluent Bit contains expected `[INPUT]` block

```
jschulz@testtcp:~$ sudo service google-cloud-ops-agent* status
# Confirm all active
jschulz@testtcp:~$ cat /var/run/google-cloud-ops-agent-fluent-bit/fluent_bit_main.conf
...
[INPUT]
    Format        json
    Listen        0.0.0.0
    Mem_Buf_Limit 10M
    Name          tcp
    Port          5170
    Tag           tcp_pipeline.tcp_logs
    storage.type  filesystem
...
```

#### 6. Send test log via TCP
```
jschulz@testtcp:~$ sudo apt-get install netcat
jschulz@testtcp:~$ echo '{"key":"value"}' | netcat 0.0.0.0 5170 -q 1
```

#### 7. Verify log ingestion via Logs Explorer

image

### CODE REVIEW

File: confgenerator/testdata/valid/linux/logging-receiver_tcp_omitting_optional_parameters/input.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+logging:
+  receivers:
+    test_tcp_log_source_id:
+      type: tcp
+  service:
+    pipelines:
+      default_pipeline:
```
[the remark this answers is not in the document]
Comment (jkschulz) [reply confgenerator/testdata/valid/linux/logging-receiver_tcp_omitting_optional_parameters/input.yaml]: ah good point. updated test files to use their own pipeline and improve naming

File: confgenerator/logging_receivers.go [hunk 2]
```diff
@@ -152,6 +152,53 @@ func init() {
...
+	ListenPort uint16 `yaml:"listen_port,omitempty"`
+}
+
+func (r LoggingReceiverTCP) Type() string {
+	return "tcp"
+}
+
+func (r LoggingReceiverTCP) Components(tag string) []fluentbit.Component {
+	if r.ListenHost == "" {
+		r.ListenHost = "0.0.0.0"
```
Comment (qingling128) [inline confgenerator/logging_receivers.go]: Had some offline discussion about the default value here.

google-fluentd defaults to `127.0.0.1`: https://github.com/GoogleCloudPlatform/fluentd-catch-all-config/blob/bf6ff8ee98a0d0fedf88f2e95113fc6c172cc986/configs/config.d/syslog_endpoint.conf#L5

We'd like to ensure the default / out of the box experience limits the access as much as possible.

File: confgenerator/logging_receivers.go [hunk 3]
```diff
@@ -152,6 +152,53 @@ func init() {
...
+
+	return []fluentbit.Component{{
+		Kind: "INPUT",
+		Config: map[string]string{
+			// https://docs.fluentbit.io/manual/pipeline/inputs/tcp
+			"Name":   "tcp",
+			"Tag":    tag,
+			"Listen": r.ListenHost,
+			"Port":   fmt.Sprintf("%d", r.ListenPort),
+			"Format": "json",
```
Comment (qingling128) [inline confgenerator/logging_receivers.go:183]: Just for the record.

There was some discussion about defaulting to `json` or `new-line split format`. Spot checking Fluentd users' tcp plugin configuration, `format json` seems a bit more popular than `format none`: https://github.com/fluent/fluentd/issues?q=is%3Aissue+%22%40type+tcp%22+

I think it's fine to default to `json` for now, and add a flag for users to control the format later if users need that feature.

**Review by qingling128 — APPROVED** [review]
LGTM

**Comment by quentinmit** [comment]
I explicitly do not approve this PR.

**Pull request merged**